README.md
# Sidekiq [![Build Status](https://secure.travis-ci.org/wallgig/chef-sidekiq.png)](http://travis-ci.org/wallgig/chef-sidekiq)

Chef cookbook for [sidekiq](http://sidekiq.org/).

# Requirements

## Chef

Tested on chef 11

## Platforms

* Debian 7+
* Ubuntu 13.10+

## Cookbooks

The following cookbooks are required:

* [runit](http://github.com/hw-cookbooks/runit)
* [logrotate](https://github.com/stevendanna/logrotate)

## Resources/Providers

### sidekiq

This generates a sidekiq configuration and creates a [runit](http://smarden.org/runit/) service. This cookbooks expects that you are deploying with
capistrano, but should be flexible enough to tune for whatever you need.

### Actions

* :create create a named sidekiq configuration, and service.
* :delete disable a named sidekiq service, and deletes the sidekiq directory.

### Examples

```ruby
sidekiq 'example.com' do
  concurrency 2
  processes 2
  queues 'job-queue' => 5, 'other-queue' => 1
end
```

```ruby
sidekiq 'example.com' do
  concurrency 2
  processes 2
  queues 'job-queue' => 5, 'other-queue' => 1
  directory '/srv/www/myapp'
  sidekiq_dir '/srv/www/myapp/sidekiq'
end
```

```ruby
sidekiq 'example.com' do
  action :delete
end
```

### Attributes
<table>
  <thead>
    <tr>
      <th>Attribute</th>
      <th>Description</th>
      <th>Default Value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>name</td>
      <td><b>Name attribute:</b> The name of the sidekiq instance.</td>
      <td><code>nil</code></td>
    </tr>
    <tr>
      <td>queues</td>
      <td>A hash of sidekiq queues</td>
      <td><code>nil</code></td>
    </tr>
    <tr>
      <td>verbose</td>
      <td>Should the sidekiq daemon be verbose, useful for debugging.</td>
      <td><code>false</code></td>
    </tr>
    <tr>
      <td>concurrency</td>
      <td>Number of concurrent sidekiq processes</td>
      <td><code>25</code></td>
    </tr>
    <tr>
      <td>processes</td>
      <td>The number of processes</td>
      <td><code>1</code></td>
    </tr>
    <tr>
      <td>timeout</td>
      <td>Timeout for sidekiq jobs, in seconds</td>
      <td><code>30</code></td>
    </tr>
    <tr>
      <td>rails_env</td>
      <td>Your rails environment</td>
      <td><code>production</code></td>
    </tr>
    <tr>
      <td>bundle_exec</td>
      <td>Should bundle exec be used to start sidekiq</td>
      <td><code>true</code></td>
    </tr>
    <tr>
      <td>bundle_command</td>
      <td>Allows you to set the path to your bundler executable</td>
      <td><code>bundle</code></td>
    </tr>
    <tr>
      <td>logrotate</td>
      <td>Enables log rotation with logrotate</td>
      <td><code>true</code></td>
    </tr>
    <tr>
      <td>owner</td>
      <td>The user of the sidekiq process</td>
      <td><code>www-data</code></td>
    </tr>
    <tr>
      <td>group</td>
      <td>The group of the sidekiq process</td>
      <td><code>www-data</code></td>
    </tr>
   </tr>
  </tbody>
</table>

# Issues

Find a bug? Want a feature? Submit an [issue here](http://github.com/wallgig/chef-sidekiq/issues). Patches welcome!

# Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

# LICENSE & AUTHORS #

* Authors:: Greg Fitzgerald (<greg@gregf.org>)

```
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# Changelog

## v0.5.0 (March 17, 2015)

* Add option to pass bundle_command (PR #2) Thanks to [@miensol](https://github.com/miensol)

## v0.4.1 (June 05, 2014)

* Fix permissions on logfiles

## v0.4.0 (May 18, 2014)

* Add a logrotate option for rotating sidekiq logs

## v0.3.0 (May 17, 2014)

* Reload service when template is updated
* action :delete now deletes sidekiq_dir and disables the service.

## v0.2.0 (May 16, 2014)

* Fixed a bug where some of the lwrp attributes could not be changed.

## v0.1.0 (May 13, 2014)

* Initial public release


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

# This is causing some tests to appear to fail even when they are passing.
# I believe it will be fixed in the next update.
#ChefSpec::Coverage.start!


spec/unit/recipes/default_spec.rb
require 'spec_helper'

describe 'sidekiq::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }
end


test/fixtures/cookbooks/sidekiq-test/metadata.rb
name 'sidekiq-test'
maintainer       'Greg Fitzgerald'
maintainer_email 'greg@gregf.org'
license          'Apache2'
description      'Tests for the sidekiq LWRP'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.1.0'

depends 'ruby_install'
depends 'sidekiq'


resources/default.rb
# Author:: Greg Fitzgerald (greg@gregf.org)
# Copyright:: Copyright (c) 2014 Greg Fitzgerald
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

actions :create, :delete
default_action :create

attribute :name, :kind_of => String, :name_attribute => true, :required => true
attribute :template, :kind_of => String, :default => 'sidekiq.yml.erb'
attribute :cookbook, :kind_of => String, :default => 'sidekiq'
attribute :queues, :kind_of => Hash, :required => true
attribute :verbose, :kind_of => [TrueClass, FalseClass], :default => false
attribute :concurrency, :kind_of => Integer, :default => 25
attribute :processes, :kind_of => Integer, :default => 1
attribute :timeout, :kind_of => Integer, :default => 30
attribute :rails_env, :kind_of => String, :default => 'production'
attribute :bundle_exec, :kind_of => [TrueClass, FalseClass], :default => true
attribute :bundle_command, :kind_of => String, :default => 'bundle'
attribute :logrotate, :kind_of => [TrueClass, FalseClass], :default => true
attribute :owner, :regex => Chef::Config[:user_valid_regex], :default => 'www-data'
attribute :group, :regex => Chef::Config[:group_valid_regex], :default => 'www-data'
attribute :directory, :kind_of => [String, NilClass], :default => nil
attribute :working_dir, :kind_of => [String, NilClass], :default => nil
attribute :sidekiq_dir, :kind_of => [String, NilClass], :default => nil
attribute :sidekiq_config, :kind_of => [String, NilClass], :default => nil
attribute :pidfile, :kind_of => [String, NilClass], :default => nil
attribute :logfile, :kind_of => [String, NilClass], :default => nil

def initialize(*args)
  super
  @action = :create
end
